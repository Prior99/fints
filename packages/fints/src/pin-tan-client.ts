import { Client } from "./client";
import { Dialog, DialogConfig } from "./dialog";
import { Request } from "./request";
import { HttpConnection } from "./http-connection";
import { Segment } from "./segments";
import { Connection } from "./types";
import { PRODUCT_NAME } from "./constants";

/**
 * Set of options needed to construct a `PinTanClient`.
 */
export interface PinTanClientConfig {
    /**
     * The fints product identification.
     */
    productId?: string;
    /**
     * The banks identification number (Bankleitzahl).
     */
    blz: string;
    /**
     * The username or identification number.
     */
    name: string;
    /**
     * The pin code or password used for authenticating with the fints server.
     */
    pin: string;
    /**
     * The URL to reach the server at.
     */
    url: string;
    /**
     * If set to `true`, will log all requests performed and responses received.
     */
    debug?: boolean;
}

export class PinTanClient extends Client {
    /**
     * Connection used to reach the server.
     */
    private connection: Connection;
    /**
     * Configuration for connecting and authenticating.
     */
    protected config: PinTanClientConfig;

    constructor(config: PinTanClientConfig) {
        super();
        this.config = config;
        const { url, debug } = config;
        this.connection = new HttpConnection({ url, debug });
    }

    public createDialog(dialogConfig?: DialogConfig) {
        const { blz, name, pin, productId = PRODUCT_NAME } = this.config;
        const { connection } = this;
        return new Dialog(dialogConfig ? dialogConfig : { blz, name, pin, systemId: "0", productId }, connection);
    }

    public createRequest(dialog: Dialog, segments: Segment<any>[], tan?: string) {
        const { blz, name, pin } = this.config;
        const { systemId, dialogId, msgNo, tanMethods } = dialog;
        return new Request({ blz, name, pin, systemId, dialogId, msgNo, segments, tanMethods, tan });
    }
}
