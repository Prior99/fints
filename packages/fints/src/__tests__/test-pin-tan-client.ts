import { format } from "date-fns";
import { readFileSync } from "fs";
import fetchMock from "fetch-mock";
import { PinTanClient } from "../pin-tan-client";
import { encodeBase64, decodeBase64 } from "../utils";
import { Format } from "../format";

const url = "https://example.com/fints";
const name = "test1";
const pin = "12345";
const blz = "12345678";
const productId = "fints";

let client: PinTanClient;

beforeEach(() => {
    jest.spyOn(Format, "date").mockImplementation(date => date ? format(date, "HHMMss") : "20180101");
    jest.spyOn(Format, "time").mockImplementation(time => time ? format(time, "HHMMss") : "120000");
    jest.spyOn(Math, "random").mockReturnValue(0.5);
    client = new PinTanClient({ blz, name, pin, url, productId });
});

test("accounts", async () => {
    const responseFixtures: string[] = JSON.parse(readFileSync(`${__dirname}/fixture-accounts.json`, "utf8"));
    let responseNo = 0;
    const mock = fetchMock.post(url, () => {
        const response = encodeBase64(responseFixtures[responseNo]);
        responseNo++;
        return response;
    });

    const result = await client.accounts();
    expect(result).toMatchSnapshot();
    const calls = (mock.calls() as any).map((call: any) => decodeBase64(String(call[1].body)));
    expect(calls).toMatchSnapshot();
    mock.restore();
});

test("statements", async () => {
    const responseFixtures: string[] = JSON.parse(readFileSync(`${__dirname}/fixture-statements.json`, "utf8"));
    let responseNo = 0;
    const mock = fetchMock.post(url, () => {
        const response = encodeBase64(responseFixtures[responseNo]);
        responseNo++;
        return response;
    });
    const account = {
        accountNumber: "2",
        bic: "GENODE00TES",
        blz: "12345678",
        iban: "DE111234567800000002",
        subAccount: "",
    };
    const result = await client.statements(account, new Date("2018-01-01T12:00:00Z"), new Date("2018-10-01T12:00:00Z"));
    expect(result).toMatchSnapshot();
    const calls = (mock.calls() as any).map((call: any) => decodeBase64(String(call[1].body)));
    expect(calls).toMatchSnapshot();
    mock.restore();
});

test("balance", async () => {
    const responseFixtures: string[] = JSON.parse(readFileSync(`${__dirname}/fixture-balance.json`, "utf8"));
    let responseNo = 0;
    const mock = fetchMock.post(url, () => {
        const response = encodeBase64(responseFixtures[responseNo]);
        responseNo++;
        return response;
    });
    const account = {
        accountNumber: "2",
        bic: "GENODE00TES",
        blz: "12346789",
        iban: "DE111234567800000002",
        subAccount: "",
    };
    const result = await client.balance(account);
    expect(result).toMatchSnapshot();
    const calls = (mock.calls() as any).map((call: any) => decodeBase64(String(call[1].body)));
    expect(calls).toMatchSnapshot();
    mock.restore();
});

test("standingOrders", async () => {
    let responseFixtures: string[] = JSON.parse(readFileSync(`${__dirname}/fixture-standingOrders.json`, "utf8"));
    let responseNo = 0;
    const mock = fetchMock.post(url, () => {
        const response = encodeBase64(responseFixtures[responseNo]);
        responseNo++;
        return response;
    });
    const account = {
        accountNumber: "2",
        bic: "DEUTDEFF500",
        blz: "12346789",
        iban: "DE27100777770209299700",
        subAccount: "",
    };
    const result = await client.standingOrders(account);
    expect(result).toMatchSnapshot();
    const calls = (mock.calls() as any).map((call: any) => decodeBase64(String(call[1].body)));
    expect(calls).toMatchSnapshot();
    mock.restore();
});
