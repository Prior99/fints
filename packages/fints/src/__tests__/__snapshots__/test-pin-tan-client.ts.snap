// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts 1`] = `
[
  {
    "accountNumber": "1",
    "bic": "GENODE00TES",
    "blz": "12345678",
    "iban": "DE111234567800000001",
    "subAccount": "",
  },
  {
    "accountNumber": "2",
    "bic": "GENODE00TES",
    "blz": "12345678",
    "iban": "DE111234567800000002",
    "subAccount": "",
  },
]
`;

exports[`accounts 2`] = `
[
  "HNHBK:1:3+000000000352+300+0+1'HNVSK:998:3+PIN:1+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@194@HNSHK:2:4+PIN:1+999+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HKSYN:5:3+0'HNSHA:6:2+5499999++12345''HNHBS:7:1+1'",
  "HNHBK:1:3+000000000369+300+DIA_10000002+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000002'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000340+300+0+1'HNVSK:998:3+PIN:2+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@182@HNSHK:2:4+PIN:2+942+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HNSHA:5:2+5499999++12345''HNHBS:6:1+1'",
  "HNHBK:1:3+000000000357+300+DIA_10000003+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@161@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKSPA:3:1+'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000369+300+DIA_10000003+3'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000003'HNSHA:4:2+5499999++12345''HNHBS:5:1+3'",
]
`;

exports[`balance 1`] = `
{
  "account": {
    "accountNumber": "2",
    "bic": "GENODE00TES",
    "blz": "12346789",
    "iban": "DE111234567800000002",
    "subAccount": "",
  },
  "availableBalance": 104.36,
  "bookedBalance": 4.36,
  "creditLimit": 100,
  "currency": "EUR",
  "pendingBalance": 0,
  "productName": "Normalsparen",
}
`;

exports[`balance 2`] = `
[
  "HNHBK:1:3+000000000352+300+0+1'HNVSK:998:3+PIN:1+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@194@HNSHK:2:4+PIN:1+999+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HKSYN:5:3+0'HNSHA:6:2+5499999++12345''HNHBS:7:1+1'",
  "HNHBK:1:3+000000000369+300+DIA_10000000+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000000'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000340+300+0+1'HNVSK:998:3+PIN:2+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@182@HNSHK:2:4+PIN:2+942+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HNSHA:5:2+5499999++12345''HNHBS:6:1+1'",
  "HNHBK:1:3+000000000409+300+DIA_10000001+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@213@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKSAL:3:7+DE111234567800000002:GENODE00TES:2::280:12346789+N++'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000369+300+DIA_10000001+3'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000001'HNSHA:4:2+5499999++12345''HNHBS:5:1+3'",
]
`;

exports[`standingOrders 1`] = `[]`;

exports[`standingOrders 2`] = `
[
  "HNHBK:1:3+000000000352+300+0+1'HNVSK:998:3+PIN:1+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@194@HNSHK:2:4+PIN:1+999+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HKSYN:5:3+0'HNSHA:6:2+5499999++12345''HNHBS:7:1+1'",
  "HNHBK:1:3+000000000369+300+DIA_10000000+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000000'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000340+300+0+1'HNVSK:998:3+PIN:2+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@182@HNSHK:2:4+PIN:2+942+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HNSHA:5:2+5499999++12345''HNHBS:6:1+1'",
  "HNHBK:1:3+000000000395+300+DIA_10000001+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@199@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKCDB:3:1+DE27100777770209299700:DEUTDEFF500++++'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000369+300+DIA_10000001+3'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000001'HNSHA:4:2+5499999++12345''HNHBS:5:1+3'",
]
`;

exports[`statements 1`] = `
[
  {
    "accountId": "12345678/0000000001",
    "closingBalance": {
      "currency": "EUR",
      "date": "2015-01-01",
      "isCredit": true,
      "value": 1223.57,
    },
    "number": "0",
    "openingBalance": {
      "currency": "EUR",
      "date": "2015-01-01",
      "isCredit": true,
      "value": 1041.23,
    },
    "referenceNumber": "STARTUMS",
    "transactions": [
      {
        "amount": 182.34,
        "bankReference": "",
        "code": "NMSC",
        "currency": "EUR",
        "customerReference": "",
        "description": "051?00UEBERWEISG?10931?20Ihre Kontonummer 0000001234?21/Test Ueberweisung 1?22n WS EREF: 1100011011 IBAN:?23 DE1100000100000001234 BIC?24: GENODE11 ?1011010100?31?32Bank",
        "descriptionStructured": {
          "bic": undefined,
          "iban": "",
          "name": "Bank",
          "primaNota": "11010100",
          "reference": {
            "raw": "Ihre Kontonummer 0000001234/Test Ueberweisung 1n WS EREF: 1100011011 IBAN: DE1100000100000001234 BIC: GENODE11 ",
          },
          "text": "UEBERWEISG",
        },
        "entryDate": "",
        "fundsCode": "",
        "id": "d9c7f240343c7e5951babcd798b4e5e7",
        "isCredit": true,
        "isExpense": true,
        "valueDate": "2015-01-01",
      },
    ],
  },
  {
    "accountId": "12345678/0000000001",
    "closingBalance": {
      "currency": "EUR",
      "date": "2015-01-01",
      "isCredit": true,
      "value": 1423.6,
    },
    "number": "0",
    "openingBalance": {
      "currency": "EUR",
      "date": "2015-03-01",
      "isCredit": true,
      "value": 1223.57,
    },
    "referenceNumber": "STARTUMS",
    "transactions": [
      {
        "amount": 100.03,
        "bankReference": "",
        "code": "NMSC",
        "currency": "EUR",
        "customerReference": "",
        "description": "051?00UEBERWEISG?10931?20Ihre Kontonummer 0000001234?21/Test Ueberweisung 2?22n WS EREF: 1100011011 IBAN:?23 DE1100000100000001234 BIC?24: GENODE11 ?1011010100?31?32Bank",
        "descriptionStructured": {
          "bic": undefined,
          "iban": "",
          "name": "Bank",
          "primaNota": "11010100",
          "reference": {
            "raw": "Ihre Kontonummer 0000001234/Test Ueberweisung 2n WS EREF: 1100011011 IBAN: DE1100000100000001234 BIC: GENODE11 ",
          },
          "text": "UEBERWEISG",
        },
        "entryDate": "",
        "fundsCode": "",
        "id": "d7c233aba577561f87ce2be780ef10ec",
        "isCredit": true,
        "isExpense": true,
        "valueDate": "2015-03-01",
      },
      {
        "amount": 100,
        "bankReference": "",
        "code": "NMSC",
        "currency": "EUR",
        "customerReference": "",
        "description": "051?00UEBERWEISG?10931?20Ihre Kontonummer 0000001234?21/Test Ueberweisung 3?22n WS EREF: 1100011011 IBAN:?23 DE1100000100000001234 BIC?24: GENODE11 ?1011010100?31?32Bank",
        "descriptionStructured": {
          "bic": undefined,
          "iban": "",
          "name": "Bank",
          "primaNota": "11010100",
          "reference": {
            "raw": "Ihre Kontonummer 0000001234/Test Ueberweisung 3n WS EREF: 1100011011 IBAN: DE1100000100000001234 BIC: GENODE11 ",
          },
          "text": "UEBERWEISG",
        },
        "entryDate": "",
        "fundsCode": "",
        "id": "908ed8c5837b54fb406ea4f1137a6629",
        "isCredit": true,
        "isExpense": true,
        "valueDate": "2015-03-01",
      },
    ],
  },
]
`;

exports[`statements 2`] = `
[
  "HNHBK:1:3+000000000352+300+0+1'HNVSK:998:3+PIN:1+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@194@HNSHK:2:4+PIN:1+999+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HKSYN:5:3+0'HNSHA:6:2+5499999++12345''HNHBS:7:1+1'",
  "HNHBK:1:3+000000000369+300+DIA_10000006+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000006'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000340+300+0+1'HNVSK:998:3+PIN:2+998+1+1::0+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@182@HNSHK:2:4+PIN:2+942+5499999+1+1+1::0+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKIDN:3:2+280:12345678+test1+0+1'HKVVB:4:3+0+0+0+fints+0.1'HNSHA:5:2+5499999++12345''HNHBS:6:1+1'",
  "HNHBK:1:3+000000000423+300+DIA_10000007+2'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@227@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKKAZ:3:7+DE111234567800000002:GENODE00TES:2::280:12345678+N+120100+121000++'HNSHA:4:2+5499999++12345''HNHBS:5:1+2'",
  "HNHBK:1:3+000000000369+300+DIA_10000007+3'HNVSK:998:3+PIN:2+998+1+1::DDDA10000000000000000000000A+1:20180101:120000+2:2:13:@8@00000000:5:1+280:12345678:test1:S:0:0+0'HNVSD:999:1+@173@HNSHK:2:4+PIN:2+942+5499999+1+1+1::DDDA10000000000000000000000A+1+1:20180101:120000+1:999:1+6:10:16+280:12345678:test1:S:0:0'HKEND:3:1+DIA_10000007'HNSHA:4:2+5499999++12345''HNHBS:5:1+3'",
]
`;
